### PR TITLE
refactor: [M3-7380] - Use `volumes/types` endpoint for pricing data

### PR DESCRIPTION
**Pull request by mjac0bs**
## Description 📝
API is now returning pricing data through the `/volumes/types` endpoint. 

In this PR, prices can be dynamically displayed for base region pricing and DC-specific region pricing, using the prices returned in the new endpoint, for Volumes.

## Changes  🔄
- Created the necessary api-v4 endpoint, query, and interface to use the volumes/types endpoint.
- Modified `getDCSpecificPriceByType` to optionally accept a size param, if size is a factor of price and updated util unit test accordingly.
- Created the necessary factory and endpoint in the MSW to support a mocked response to the API endpoint.
- Added UX for error handling in the unlikely event that pricing cannot be calculated: disable create button and display tooltip error. 
- Added loading state for display prices.
- Updated `create-volume-smoke.spec.ts` to test create volume error flows.

Note: I did *not* update the resize or clone Cypress tests to test those error flows. These existing specs run slowly since they have to create a linode and I'm not sure it's worth it for an edge-case like this, so held off doing that and will talk to Joe when he returns.

## Preview 📷
The **only** visual changes should be loading state for pricing and the disabled Create Volume, Resize Volume, and Clone Volume buttons *if* the price is invalid (e.g. undefined).
| Flow | Screenshot |
| --- | --- |
| Create Volume Landing Error State | Screenshot 2024-04-12 at 12 17 57 PM |
| Create Volume from Linode Error State | Screenshot 2024-04-12 at 12 18 18 PM |
| Resize Volume Error State | Screenshot 2024-04-12 at 12 18 43 PM |
| Clone Volume Error State | Screenshot 2024-04-12 at 10 41 33 AM |

### Prerequisites
(How to setup test environment)
- Check out this PR and `yarn up`.
- All new `types` endpoints are all in prod, so use the prod environment.
- Have an existing linode or create one. 

### Verification steps
(How to verify changes)
- Go to http://localhost:3000/volumes/create.
- Fill out the form.
- Verify price calculation is correct by testing different options for the Region and comparing with prod. 
- Confirm:
   - There are no regressions in decimal point value or calculations displayed in the UI.
   - $0 is a valid region price if returned from the API, the user sees $0 pricing, and a volume can be created. (Update the mock data to test this.)
   - A user can still create a Volume.
   - If price cannot be calculated (e.g. it is undefined), the Summary displays the unknown price ($--.--) string, the Create button is disabled, and a tooltip displays on hover. (Blocking network requests is one method.)
   - When a region is not yet selected, the Create button is still enabled. (Pricing can't be calculated yet; we don't know if the price is invalid.)
   - Loading state functions as expected. 
- Additionally, confirm the above when creating or resizing a volume _from a linode_:
   - via Linode Details > Storage > Create Volume
   -  via Linode Details > Storage > Clone (the volume you created in the last step)
   - via Linode Details > Storage > Resize (the volume you created)

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [x] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [x] 🧪 Providing/Improving test coverage
- [x] 🔐 Removing all sensitive information from the code and PR description
- [ ] 🚩 Using a feature flag to protect the release
- [x] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

### CODE REVIEW

File: packages/manager/src/features/Volumes/VolumeDrawer/PricePanel.tsx [hunk 1]
```diff
@@ -32,7 +36,11 @@ export const PricePanel = ({ currentSize, regionId, value }: Props) => {
 
   return (
     <Box marginTop={4}>
-      <DisplayPrice interval="mo" price={Number(price)} />
+      {isLoading ? (
```
Comment (mjac0bs) [inline packages/manager/src/features/Volumes/VolumeDrawer/PricePanel.tsx:39]: Added loading state for the display price.

https://github.com/linode/manager/assets/114685994/4723782a-ac73-45c1-9962-2dc10387bb7a

File: packages/manager/src/features/Volumes/VolumeDrawer/SizeField.tsx [hunk 2]
```diff
@@ -80,7 +85,12 @@ export const SizeField = (props: Props) => {
     </FormHelperText>
   );
 
-  const dynamicPricingHelperText = !resize && !isFromLinode && (
+  // The price display is only visible next to the size field on the Volumes Create page.
```
Comment (mjac0bs) [inline packages/manager/src/features/Volumes/VolumeDrawer/SizeField.tsx:88]: Did a little bit of renaming for clarity here and also added loading state to the price display for volumes create page to improve the slow network or error experience.

https://github.com/linode/manager/assets/114685994/17c4842e-e8ba-4b3e-92d9-5edd7e60dbfb

File: packages/manager/src/utilities/pricing/dynamicPricing.test.ts [hunk 3]
```diff
@@ -81,6 +75,16 @@ describe('getDCSpecificPricingByType', () => {
     ).toBe('14.00');
   });
 
+  it('calculates dynamic pricing for a volume based on size', () => {
```
Comment (mjac0bs) [inline packages/manager/src/utilities/pricing/dynamicPricing.test.ts:78]: We can remove `getDynamicVolumePrice` entirely now that we have price data in the `/types` endpoint; we just have to allow an optional size to be passed into `getDCSpecificPriceByType`. Test coverage is updated here accordingly.

**Review by abailly-akamai — APPROVED** [review]
Nice and clean! 

Code is good and consistent ✅ 
UI Price display, error and loading handling look great ✅ 

Left a couple comments for consideration

File: packages/manager/.changeset/pr-10376-tech-stories-1712948758226.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"@linode/manager": Tech Stories
```
[the remark this answers is not in the document]
Comment (bnussman-akamai) [reply packages/manager/.changeset/pr-10376-tech-stories-1712948758226.md:2]: I think I agree. This change is somewhat customer-facing

**Review by bnussman-akamai — APPROVED** [review]
Awesome work! 🎉  Didn't spot any issues 👁️

File: packages/manager/src/features/Volumes/VolumeDrawer/LinodeVolumeCreateForm.tsx [hunk 5]
```diff
@@ -56,10 +60,12 @@ export const LinodeVolumeCreateForm = (props: Props) => {
   const { data: profile } = useProfile();
   const { data: grants } = useGrants();
   const { mutateAsync: createVolume } = useCreateVolumeMutation();
+  const { data: types, isError, isLoading } = useVolumeTypesQuery();
 
   const { checkForNewEvents } = useEventsPollingActions();
 
-  const disabled = profile?.restricted && !grants?.global.add_volumes;
+  const isReadOnly = profile?.restricted && !grants?.global.add_volumes;
```
[the remark this answers is not in the document]
Comment (mjac0bs) [reply packages/manager/src/features/Volumes/VolumeDrawer/LinodeVolumeCreateForm.tsx:67]: Good call, I think so. I'll update.

File: packages/manager/.changeset/pr-10376-tech-stories-1712948758226.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (mjac0bs) [reply packages/manager/.changeset/pr-10376-tech-stories-1712948758226.md]: Yeah, that's fair. I remember debating where to put the `nodebalancers/types` changeset and it ended up in Tech Stories, so I kept that consistent here, but think you're both right that it's more customer-facing that not and belongs in changed. Fixed in [4eb896a](https://github.com/linode/manager/pull/10376/commits/4eb896abcacb4ee002533e322544ec663c44990c).
